packages/discord.js/src/managers/GuildMemberManager.js

```diff
@@ -206,18 +204,76 @@ class GuildMemberManager extends CachedManager {
    */
   fetch(options) {
     if (!options) return this._fetchMany();
-    const user = this.client.users.resolveId(options);
-    if (user) return this._fetchSingle({ user, cache: true });
-    if (options.user) {
-      if (Array.isArray(options.user)) {
-        options.user = options.user.map(u => this.client.users.resolveId(u));
-        return this._fetchMany(options);
-      } else {
-        options.user = this.client.users.resolveId(options.user);
-      }
-      if (!options.limit && !options.withPresences) return this._fetchSingle(options);
+    const { user, cache, force } = options;
+
+    const {
+      client: { users },
+    } = this;
+
+    const resolvedUser = users.resolveId(user ?? options);
+    if (resolvedUser) return this._fetchSingle({ user: resolvedUser, cache, force });
+    const resolvedUsers = user?.map(_user => users.resolveId(_user));
+    return this._fetchMany({ ...options, users: resolvedUsers });
+  }
+
+  async _fetchSingle({ user, cache, force = false }) {
+    if (!force) {
+      const existing = this.cache.get(user);
+      if (existing && !existing.partial) return existing;
     }
-    return this._fetchMany(options);
+
+    const data = await this.client.rest.get(Routes.guildMember(this.guild.id, user));
+    return this._add(data, cache);
+  }
+
+  _fetchMany({
+    limit = 0,
+    withPresences: presences,
+    users,
+    query,
+    time = 120e3,
+    nonce = DiscordSnowflake.generate().toString(),
+  } = {}) {
+    return new Promise((resolve, reject) => {
+      if (!query && !users) query = '';
+      if (nonce.length > 32) throw new DiscordjsRangeError(ErrorCodes.MemberFetchNonceLength);
+      this.guild.shard.send({
+        op: GatewayOpcodes.RequestGuildMembers,
+        d: {
+          guild_id: this.guild.id,
+          presences,
+          user_ids: users,
+          query,
+          nonce,
+          limit,
+        },
+      });
+      const fetchedMembers = new Collection();
+      let i = 0;
+      const handler = (members, _, chunk) => {
+        timeout.refresh();
+        if (chunk.nonce !== nonce) return;
```

Comment:
I don't know if we should refresh the timeout on every event, rather than on every event sent with the same nonce.

Reply:
I believe it should be confined at best to prevent unexpected behaviour. The `nonce` check has been prioritised with f06f747fad7f729f901eac0c7e8da0aed7903a66.